File: Tracer.js
/**
 * @module Tracer
 * @description A simplified, secure tracing implementation using native Node.js libraries.
 */

const crypto = require('crypto');

/**
 * @class Tracer
 * @description A basic tracer for creating and managing spans.
 */
class Tracer {
  /**
   * @constructor
   * @param {string} serviceName - The name of the service being traced.
   */
  constructor(serviceName) {
    /** @private */
    this.serviceName = serviceName;
    /** @private */
    this.traces = {};
  }

  /**
   * @private
   * @function generateSecureId
   * @description Generates a cryptographically secure random ID.
   * @returns {string} - A hexadecimal string representing the ID.
   */
  generateSecureId() {
    return crypto.randomBytes(16).toString('hex');
  }

  /**
   * @function startSpan
   * @description Starts a new span.
   * @param {string} name - The name of the span.
   * @param {string} [parentSpanId] - The ID of the parent span (if any).
   * @returns {object} - The newly created span object.
   */
  startSpan(name, parentSpanId = null) {
    const traceId = parentSpanId ? this.traces[parentSpanId][0].traceId : this.generateSecureId();
    const spanId = this.generateSecureId();
    const startTime = Date.now();

    const span = {
      traceId,
      spanId,
      parentSpanId,
      name,
      startTime,
      endTime: null,
      attributes: {},
      status: 'UNSET',
    };

    if (!this.traces[traceId]) {
      this.traces[traceId] = [];
    }
    this.traces[traceId].push(span);

    return span;
  }

  /**
   * @function endSpan
   * @description Ends a span and sets its status and attributes.
   * @param {object} span - The span object to end.
   * @param {string} [status='OK'] - The status of the span ('OK' or 'ERROR').
   * @param {object} [attributes={}] - Additional attributes to add to the span.
   */
  endSpan(span, status = 'OK', attributes = {}) {
    span.endTime = Date.now();
    span.status = status;
    span.attributes = { ...span.attributes, ...attributes };
  }

  /**
   * @function getTrace
   * @description Retrieves all spans for a given trace ID.
   * @param {string} traceId - The ID of the trace.
   * @returns {array} - An array of span objects.
   */
  getTrace(traceId) {
    return this.traces[traceId];
  }

  /**
   * @function exportTraces
   * @description Exports the collected traces (simulated in this example).
   */
  exportTraces() {
    console.log('Exporting traces...');
    console.log(JSON.stringify(this.traces, null, 2));
  }
}

module.exports = { Tracer };

File: tracer.go
package tracer

import (
        "context"
        "crypto/rand"
        "encoding/hex"
        "fmt"
        "log"
        "time"

        "github.com/dash-xd/gospace/internal/axiomlogger"
)

type Span struct {
        TraceID      string            `json:"traceId"`
        SpanID       string            `json:"spanId"`
        ParentSpanID string            `json:"parentSpanId,omitempty"`
        Name         string            `json:"name"`
        StartTime    time.Time         `json:"startTime"`
        EndTime      *time.Time        `json:"endTime,omitempty"`
        Attributes   map[string]string `json:"attributes"`
        Status       string            `json:"status"`
}

type Tracer struct {
        ServiceName string
        Logger      *axiomlogger.Logger
}

func NewTracer(serviceName string, logger *axiomlogger.Logger) *Tracer {
        return &Tracer{
                ServiceName: serviceName,
                Logger:      logger,
        }
}

func (t *Tracer) generateSecureID() string {
        b := make([]byte, 16)
        _, err := rand.Read(b)
        if err != nil {
                log.Fatalf("Error generating random bytes: %v", err)
        }
        return hex.EncodeToString(b)
}

func (t *Tracer) StartSpan(ctx context.Context, name string) (context.Context, *Span) {
        var parentSpanID *string
        if parentSpan, ok := ctx.Value("span").(*Span); ok {
                parentSpanIDPtr := parentSpan.SpanID
                parentSpanID = &parentSpanIDPtr
        }

        traceID := ""
        if parentSpanID != nil {
                traceID = ctx.Value("traceID").(string)
        } else {
                traceID = t.generateSecureID()
        }

        spanID := t.generateSecureID()
        startTime := time.Now()

        span := &Span{
                TraceID:      traceID,
                SpanID:       spanID,
                ParentSpanID: "",
                Name:         name,
                StartTime:    startTime,
                Attributes:   make(map[string]string),
                Status:       "UNSET",
        }

        if parentSpanID != nil {
                span.ParentSpanID = *parentSpanID
        }

        newCtx := context.WithValue(ctx, "span", span)
        newCtx = context.WithValue(newCtx, "traceID", traceID)

        return newCtx, span
}

func (t *Tracer) EndSpan(ctx context.Context, span *Span, status string, attributes map[string]string) {
        endTime := time.Now()
        span.EndTime = &endTime
        span.Status = status
        for k, v := range attributes {
                span.Attributes[k] = v
        }

        data := map[string]interface{}{
                "traceId":      span.TraceID,
                "spanId":       span.SpanID,
                "parentSpanId": span.ParentSpanID,
                "name":         span.Name,
                "startTime":    span.StartTime,
                "endTime":      span.EndTime,
                "attributes":   span.Attributes,
                "status":       span.Status,
        }
        if err := t.Logger.LogData(data); err != nil {
                log.Printf("Error logging to Axiom: %v", err)
        }
}
